src/gnn_tracking/metrics/losses.py

```diff
@@ -226,21 +227,121 @@ def _condensation_loss(
     }
 
 
+def _first_occurrences(input_array: T):
+    # Function that determines the first occurrence of each unique element in a 1D array
+    return torch.tensor(
+        np.unique(input_array.cpu(), return_index=True)[1], device=input_array.device
+    )
+
+
+def _square_distances(edges: T, positions: T):
+    # Function that calculates the squared distances between two sets of points
+    return torch.sum((positions[edges[0]] - positions[edges[1]]) ** 2, dim=-1)
+
+
+def _fast_condensation_loss(
+    *,
+    beta: T,
+    x: T,
+    particle_id: T,
+    q_min: float,
+    mask: T,
+    radius_threshold: float,
+    max_num_neighbors: int,
+) -> dict[str, T]:
+    """Extracted function for condensation loss. See `PotentialLoss` for details."""
+    # For better readability, variables that are only relevant in one "block"
+    # are prefixed with an underscore
+
+    # -- 1. Determine indices of condensation points (CPs) and q --
+    _sorted_indices = torch.argsort(beta, descending=True)
+    _pids_sorted = particle_id[_sorted_indices]
+    _alphas = _sorted_indices[_first_occurrences(_pids_sorted)]
+    # Index of condensation points in node array
+    alphas = _alphas[particle_id[_alphas] > 0]
+    assert alphas.size()[0] > 0, "No particles found, cannot evaluate loss"
+    q = torch.arctanh(beta) ** 2 + q_min
+    assert not torch.isnan(q).any(), "q contains NaNs"
+
+    # -- 2. Edges for repulsion loss --
+    _radius_edges = radius_graph(
+        x=x, r=radius_threshold, max_num_neighbors=max_num_neighbors, loop=False
+    )
+    # Now filter out everything that doesn't include a CP or connects two hits of the
+    # same particle
+    _to_cp = torch.isin(_radius_edges[0], alphas)
+    _is_repulsive = particle_id[_to_cp[0]] != particle_id[_to_cp[1]]
+    repulsion_edges = _to_cp[:, _is_repulsive & _to_cp]
+
+    # -- 3. Edges for attractive loss --
+    # 1D array (n_nodes): 1 for CPs, 0 otherwise
+    alpha_hits_filter = torch.zeros(
+        len(particle_id), dtype=bool, device=x.device
+    ).scatter_(0, alphas, 1)
+    # indices of all non-CPs
+    non_alpha_indices = torch.arange(len(particle_id), device=x.device)[
+        ~alpha_hits_filter
+    ]
+
+    # for each non-CP hit, the index of the corresponding CP
+    alpha_indices = _alphas[
+        torch.searchsorted(particle_id[_alphas], particle_id[non_alpha_indices])
+    ]
+
+    # Insert alpha indices into their respective positions to form attraction edges
+    unmasked_attraction_edges = (
+        torch.arange(len(particle_id), device=x.device).unsqueeze(0).repeat(2, 1)
+    )
+    unmasked_attraction_edges[1, ~alpha_hits_filter] = alpha_indices
+
+    # Apply mask to attraction edges
+    noisy_attraction_edges = unmasked_attraction_edges[:, mask]
+
+    # Filter out noise edges from attraction edges
+    attraction_edges = noisy_attraction_edges[
```

Comment:
the mask will already exclude noise (`pt > 0` excludes noise if you look at the mask definition), so this isn't needed